### PR TITLE
Ensure TraitListEvent.index is always an integer

### LINKED ISSUE #540: Sync trait event is broken when setting an empty slice on a list.
Encountering a trait error when setting an empty slice (or a slice without a first index).

Code to reproduce:
```python
from traits.api import *


class Foo(HasStrictTraits):
    data = List(Int)


foo = Foo(data=[-1, -2, -3])
bar = Foo(data=[])
foo.sync_trait('data', bar, mutual=True)
print(foo.data, bar.data)
bar.data[:] = [1, 2, 3]
print(foo.data, bar.data)
```

The problem is in `HasTraits._sync_trait_items_modified`. When you have a slice like this, the `index` attribute of the trait list event is `None`, causing the logic used to form the slice to fail.

### PR DESCRIPTION
**Pull request by mdickinson**
Previously, `myobj.mylist[:2] = new_list` would generate a `TraitListEvent` with `index=None`, while `myobj.mylist[0:2] = new_list` would generate a `TraitListEvent` with `index=0`. This PR changes the `TraitListEvent` creation so that the `index` is `0` in both cases.

This is an extension of @tmreay's PR #545 that keeps the existing test but uses a fix closer to the source of the problem.

Fixes #540.

### CODE REVIEW

**Comment by mdickinson** [comment]
@rahulporuri Do you have bandwidth for a review?

@tmreay If you have time for a review, that would be welcome, but I'm assuming that you're unavailable.

File: traits/tests/test_sync_traits.py [hunk 1]
```diff
@@ -128,6 +128,33 @@ def test_sync_lists(self):
         with self.assertTraitDoesNotChange(a, "l"):
             b.l = [7, 8]
 
+    def test_sync_lists_partial_slice(self):
```
Comment (rahulporuri) [inline traits/tests/test_sync_traits.py]: I was hoping to see a CI failure for this commit. I checked out this commit specifically and I can see the failure locally. FTR, the traceback is 

```

======================================================================
ERROR: Test synchronization of list traits when there is a partial slice.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/rporuri/.edm/envs/traits-test-3.6/lib/python3.6/site-packages/traits/tests/test_sync_traits.py", line 150, in test_sync_lists_partial_slice
    a.l[:] = [7]
  File "/Users/rporuri/.edm/envs/traits-test-3.6/lib/python3.6/site-packages/traits/trait_handlers.py", line 2568, in __setitem__
    self.name_items, TraitListEvent(index, removed, values)
  File "/Users/rporuri/.edm/envs/traits-test-3.6/lib/python3.6/site-packages/traits/trait_handlers.py", line 2485, in _send_trait_items_event
    object.trait_items_event(name, event, items_event)
  File "/Users/rporuri/.edm/envs/traits-test-3.6/lib/python3.6/site-packages/traits/trait_notifiers.py", line 545, in __call__
    self.notify_listener(self, object, trait_name, old, new)
  File "/Users/rporuri/.edm/envs/traits-test-3.6/lib/python3.6/site-packages/traits/trait_notifiers.py", line 619, in _notify_method_listener
    object, trait_name, old, new, listener
  File "/Users/rporuri/.edm/envs/traits-test-3.6/lib/python3.6/site-packages/traits/trait_notifiers.py", line 599, in _dispatch_change_event
    handle_exception(object, trait_name, old, new)
  File "/Users/rporuri/.edm/envs/traits-test-3.6/lib/python3.6/site-packages/traits/trait_notifiers.py", line 180, in _handle_exception
    raise excp
  File "/Users/rporuri/.edm/envs/traits-test-3.6/lib/python3.6/site-packages/traits/trait_notifiers.py", line 591, in _dispatch_change_event
    self.dispatch(handler, *args)
  File "/Users/rporuri/.edm/envs/traits-test-3.6/lib/python3.6/site-packages/traits/trait_notifiers.py", line 553, in dispatch
    handler(*args)
  File "/Users/rporuri/.edm/envs/traits-test-3.6/lib/python3.6/site-packages/traits/has_traits.py", line 2953, in _sync_trait_items_modified
    n1 = n0 + len(event.removed)
TypeError: unsupported operand type(s) for +: 'NoneType' and 'int'

----------------------------------------------------------------------
```

File: traits/trait_handlers.py [hunk 2]
```diff
@@ -2537,7 +2537,7 @@ def __setitem__(self, key, value):
                     # In this case, we return a TraitListEvent with an
                     # index=key.start and the removed and added lists as they
                     # are.
```
Comment (rahulporuri) [inline traits/trait_handlers.py]: do we need to update the FIXME comment about this change?

File: traits/tests/test_sync_traits.py (hunk 1 above)
Comment (mdickinson) [reply traits/tests/test_sync_traits.py]: @rahulporuri I don't think we have any rule about using TDD for Traits development. Or am I misunderstanding your point? Is there something actionable here?

File: traits/trait_handlers.py (hunk 2 above)
Comment (mdickinson) [reply traits/trait_handlers.py]: That's about a different thing; nothing to do with the use of `None`. Admittedly it's not strictly true that we're using exactly `index=key.start` any more, but I think it's close enough that the comment shouldn't be confusing.

File: traits/tests/test_sync_traits.py (hunk 1 above)
Comment (rahulporuri) [reply traits/tests/test_sync_traits.py]: nothing actionable. personal preference.

File: traits/trait_handlers.py (hunk 2 above)
Comment (rahulporuri) [reply traits/trait_handlers.py]: 👍

**Pull request merged**